iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp

```diff
@@ -91,21 +91,131 @@ static void createAsyncGroups(FuncOp funcOp) {
   }
 }
 
+static Value collapseMemRefTo2D(PatternRewriter& rewriter, mlir::Location loc,
+                                Value source, MemRefType sourceType) {
+  llvm::SmallVector<ReassociationIndices> indicesArray;
+  ReassociationIndices collapasedIndices;
+  for (int i = 0; i < sourceType.getRank() - 1; i++)
+    collapasedIndices.push_back(i);
+  indicesArray.push_back(collapasedIndices);
+  indicesArray.push_back({sourceType.getRank() - 1});
+  return rewriter.create<memref::CollapseShapeOp>(loc, source, indicesArray);
+}
+
+static AffineMap get2DStridedLinearLayoutMap(MemRefType t) {
+  int64_t offset;
+  SmallVector<int64_t, 4> strides;
+  if (failed(getStridesAndOffset(t, strides, offset))) return AffineMap();
+  strides.pop_back();
+  int64_t firstStride = strides.back();
+  for (int64_t& i : strides) i = i / firstStride;
+  return makeStridedLinearLayoutMap(strides, offset, t.getContext());
+}
 namespace {
 
+struct FlattenTransferReadOp : public OpRewritePattern<vector::TransferReadOp> {
+  using OpRewritePattern<vector::TransferReadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::TransferReadOp transferReadOp,
+                                PatternRewriter& rewriter) const override {
+    auto loc = transferReadOp.getLoc();
+    Value vector = transferReadOp.vector();
+    VectorType vectorType = vector.getType().cast<VectorType>();
+    Value source = transferReadOp.source();
+    MemRefType sourceType = source.getType().dyn_cast<MemRefType>();
+    // Contiguity check is valid on tensors only.
+    if (!sourceType) return failure();
+    // Already 2D or lower nothing to do.
+    if (vectorType.getRank() < 3) return failure();
+    int numberOfNonUnitDims = 1;  // The innermost dim is always considered non
+                                  // unit as it wont be collapsed
+    int indexOfOuterMostUnitDim =
+        vectorType.getRank() - 2;  // default index for the outermost dim of
+                                   // collapsed vector in the orginal vector
+    for (int i = 0; i < vectorType.getRank() - 1; i++) {
+      if (vectorType.getShape()[i] != 1) {
+        numberOfNonUnitDims++;
+        indexOfOuterMostUnitDim = i;
+      }
+    }
+    // Bail out if 2D vector cannot be formed
+    if (numberOfNonUnitDims > 2) {
```

Comment:
nit: should be `numberOfNonUnitDims != 2`?

Reply:
I wanted to support the case like vector<1x1x1x1> to vector <1x1> see the last lit test. But let me know if its better to bail out in a case like this and I can change it.